bin/shpec

```diff
@@ -44,97 +54,138 @@ stub_command() {
+      "Expected [$2] to be blank"
+  ;;
 
   xfile_present )
-    print_result "[[ -e $2 ]]" "Expected file [$2] to exist";;
-
+    print_result "[ -e $2 ]" \
+      "Expected file [$2] to exist"
+  ;;
   xfile_absent )
-    print_result "[[ ! -e $2 ]]" "Expected file [$2] not to exist";;
-
+    print_result "[ ! -e $2 ]" \
+      "Expected file [$2] not to exist"
+  ;;
   xsymlink )
     link="$(readlink $2)"
-    print_result "[[ '$link' = '$3' ]]" "Expected [$2] to link to [$3], but got [$link]";;
-
+    print_result "[ '$link' = '$3' ]" \
+      "Expected [$2] to link to [$3], but got [$link]"
+  ;;
   xtest )
-    print_result "$2" "Expected $2 to be true";;
-
+    print_result "$2" \
+      "Expected $2 to be true"
+  ;;
   * )
     if type "$1" 2>/dev/null | grep -q 'function'; then
       matcher="$1"; shift
       $matcher "$@"
     else
       print_result false "Error: Unknown matcher [$1]"
-    fi;;
-
+    fi
+  ;;
   esac
 }
 
 print_result() {
   if eval "$1"; then
-    iecho -e "$green$assertion$norm"
+    iecho "$green$assertion$norm"
   else
-    ((failures += 1))
-    iecho -e "$red$assertion"
-    iecho -e "($2)$norm"
+    : $((failures += 1))
+    iecho "$red$assertion"
+    iecho "($2)$norm"
   fi
 }
 
-SHPEC_ROOT=${SHPEC_ROOT:-$([[ -d './shpec' ]] && echo './shpec' || echo '.')}
-
-case "$1" in
-
-  -v|--version )
-
-    echo "$VERSION";;
-
-  * )
-
-    matcher_files="$(find "$SHPEC_ROOT/matchers" -name '*.sh' 2>/dev/null)"
-
-    for matcher_file in $matcher_files; do
-      . "$matcher_file"
-    done
-
-    files="${@:-$(find $SHPEC_ROOT -name '*_shpec.sh')}"
-
-    time for file in $files; do
-      . "$file"
-    done
 
-    [ $failures -eq 0 ] && color=$green || color=$red
-    echo -e "${color}${examples} examples, ${failures} failures${norm}"
+shpec() {
+  (
+  VERSION=0.1.2posix
+  examples=0
+  failures=0
+  test_indent=0
+  red="\033[0;31m"
+  green="\033[0;32m"
+  norm="\033[0m"
+
+  SHPEC_ROOT=${SHPEC_ROOT:-$([ -d './shpec' ] && echo './shpec' || echo '.')}
+  case "$1" in
+  ( -v | --version ) echo "$VERSION"
+  ;;
+  ( * )
+      matcher_files="$(
+          find "$SHPEC_ROOT/matchers" -name '*.sh' 2>/dev/null
+      )"
+
+      for matcher_file in $matcher_files; do
+        . "$matcher_file"
+      done
+
+      if [ $# -gt 0 ] ; then
+        : \$1="'$1'" \${#1}="'${#1}'"
+        files="${@}"
+      else
+        files=$(
+          find $SHPEC_ROOT -name '*_shpec.sh'
+        )
+      fi
+
+      for file in $files; do
+        echo >& 2 "$file"
+        . "$file" || exit
+      done
+
+      [ $failures -eq 0 ] && color=$green || color=$red
+      echoe "${color}${examples} examples, ${failures} failures${norm}"
+
+      times
+      [ $failures -eq 0 ]
+      exit
+  ;;
+  esac
+  )
+}
 
-    [ $failures -eq 0 ]
+(
+  _progname=shpec
+  _pathname=$( command -v $0 )
+  _cmdname=${_pathname##*/}
+  _main=shpec
 
-esac
+  case $_progname in (${_cmdname%.sh}) $_main "$@";; esac
```

Comment:
When we run the 'shpec' script,  we want to make sure that the '$0' basename matches the name we've chosen.  ( 'shpec' )

It's a variation on idiom that I have used quite a lot.  There is a long standing 
tradition to distinguish shell 'source-files' from installed scripts by appending '.sh' to the file in the source code repository.  (Going back to sccs and make in system V.)

So we need to get rid of the directory path, and we then need to match against
either "shpec" or "shpec.sh".

So, instead of the external `basename`, I use the parameter expansion modifiers:

First, `command -v $0` will return a fully-qualified path, if there is any.
Then, ${var##*/} will strip the longest left-matching string ending in "/" from $var.
Finally, ${var%.sh} will strip the string ".sh" from the right of $var, if it's there.

Short version, if the name of the script matches 'shpec' or 'shpec.sh', we call the 'shpec' function.